### DISCUSSION TITLE
Custom Route Match

### DISCUSSION THREAD

**Discussion by nixtar** [dotnet/yarp#1781] [Q&A] [answered]
Hello,

Is it possible to implement/extend the Match property with our own logic?

In our environment we have a lot of remote sites and would like to do location based routing with yarp.

EG something like:
```json
{
  "ReverseProxy": {
    "Routes": {
      "route1" : {
        "ClusterId": "cluster1",
        "Match": {
          "SuperCustomLocations" : [ "RemoteOffice1", "RemoteOffice2"]
        },
      }
    },
    "Clusters": {
      "cluster1": {
        "Destinations": {
          "cluster1/destination1": {
            "Address": "https://example.com/"
          }
        }
      }
    }
  }
}
```

Then we have our own implementation of SuperCustomLocations with code that calculates the string value from x-forwarded-for for eg.

**Comment by nixtar** [comment, marked as answer]
Current idea is to try and implement a transform in code that injects a "Location" header into the request.
However I'm not sure that will work as I assume that to hit the transform it will have already evaluated the routes.

Maybe a middleware?

Edit:
Middleware was the answer :)

**Reply by samsp-msft** [reply]
Extending the routing logic is not a simple thing, so your approach is probably best. You will need to explicitly set the order of processing so that your middleware is called before routing (which also happens before YARP gets involved). See [routing-basics](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-6.0#routing-basics) for an example of specifying your middleware before `app.UseRouting();`

**Reply by nixtar** [reply]
Thanks @samsp-msft ❤

Doing a middleware works great.
My last attempt at this in the past I got tripped up by not setting UseRouting. My assumption at the time was that it would be called when mapping the reverse proxy. Turns out its a bit more magic then that 😄.

This solution is ideal for us as we wanted to keep the yarp config in the json as much as possible to avoid needing to recompile when we want to change the config for our location based routing. 👍


FYI what my successful PoC ended up being:

```c#
var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services.AddReverseProxy()
    .LoadFromConfig(builder.Configuration.GetSection("ReverseProxy"));

var app = builder.Build();

// Custom middleware that runs at the start of the pipeline to inject location headers
app.Use(async (context, next) =>
{
    context.Request.Headers.Add("SiteCode", "ABC");
    
    await next(context);
});

// Specify UseRouting to ensure that route matching middleware runs after the custom middleware
app.UseRouting();

app.MapReverseProxy();

app.Run();
```

```json
"ReverseProxy": {
    "Routes": {
      "route1" : {
        "ClusterId": "cluster1",
        "Match": {
          "Path": "{**catch-all}"
        }
      },
      "route2" : {
        "ClusterId": "cluster2",
        "Match": {
          "Path": "{**catch-all}",
          "Headers": [
            {
              "Name": "SiteCode",
              "Values": ["ABC", "DEF"],
              "Mode": "ExactHeader"
            }
          ]
        }
      }
    },
    "Clusters": {
      "cluster1": {
        "Destinations": {
          "destination1": {
            "Address": "http://server1"
          }
        }
      },
      "cluster2": {
        "Destinations": {
          "destination1": {
            "Address": "http://server2"
          }
        }
      }
    }
  }
```
